nav2_route/include/nav2_route/corner_smoothing.hpp

```diff
@@ -0,0 +1,134 @@
+      {(start.x - corner.x) / start_edge_length_, (start.y - corner.y) / start_edge_length_};
+      std::vector<double> end_edge_unit_tangent =
+      {(end.x - corner.x) / end_edge_length_, (end.y - corner.y) / end_edge_length_};
+
+      double bisector_x = start_edge_unit_tangent[0] + end_edge_unit_tangent[0];
+      double bisector_y = start_edge_unit_tangent[1] + end_edge_unit_tangent[1];
+      double bisector_magnitude = std::sqrt(bisector_x * bisector_x + bisector_y * bisector_y);
+
+      std::vector<double> unit_bisector =
+      {bisector_x / bisector_magnitude, bisector_y / bisector_magnitude};
+
+      start_coordinate_.x = corner.x + start_edge_unit_tangent[0] * tangent_length;
+      start_coordinate_.y = corner.y + start_edge_unit_tangent[1] * tangent_length;
+
+      end_coordinate_.x = corner.x + end_edge_unit_tangent[0] * tangent_length;
+      end_coordinate_.y = corner.y + end_edge_unit_tangent[1] * tangent_length;
+
+      double bisector_length = minimum_radius / std::sin(angle / 2);
+
+      circle_center_coordinate_.x = corner.x + unit_bisector[0] * bisector_length;
+      circle_center_coordinate_.y = corner.y + unit_bisector[1] * bisector_length;
+
+      valid_corner_ = true;
+    }
+  }
+
+  ~CornerArc() = default;
+
+  void interpolateArc(
+    const float & max_angle_resolution,
+    std::vector<geometry_msgs::msg::PoseStamped> & poses)
+  {
+    std::vector<double> r_start{start_coordinate_.x - circle_center_coordinate_.x,
+      start_coordinate_.y - circle_center_coordinate_.y};
+    std::vector<double> r_end{end_coordinate_.x - circle_center_coordinate_.x,
+      end_coordinate_.y - circle_center_coordinate_.y};
+    double cross = r_start[0] * r_end[1] - r_start[1] * r_end[0];
+    double dot = r_start[0] * r_end[0] + r_start[1] * r_end[1];
+    double signed_angle = std::atan2(cross, dot);
+    int N = std::ceil(std::abs(signed_angle) / max_angle_resolution);
+    double angle_resolution = signed_angle / N;
+
+    float x, y;
+
+    for (int i = 0; i <= N; i++) {
+      double angle = i * angle_resolution;
+      x = circle_center_coordinate_.x +
+        (r_start[0] * std::cos(angle) - r_start[1] * std::sin(angle));
+      y = circle_center_coordinate_.y +
+        (r_start[0] * std::sin(angle) + r_start[1] * std::cos(angle));
+      poses.push_back(utils::toMsg(x, y));
+    }
+  }
+
+
+  bool isCornerValid() const {return valid_corner_;}
+
+  Coordinates getCornerStart() const {return start_coordinate_;}
+
+  Coordinates getCornerEnd() const {return end_coordinate_;}
+
+protected:
+  double getAngleBetweenEdges(
+    const Coordinates & start, const Coordinates & corner,
+    const Coordinates & end)
+  {
+    double start_dx = start.x - corner.x;
+    double start_dy = start.y - corner.y;
+
+    double end_dx = end.x - corner.x;
+    double end_dy = end.y - corner.y;
+
+    double angle =
+      acos((start_dx * end_dx + start_dy * end_dy) / (start_edge_length_ * end_edge_length_));
```

Comment:
Ditto on divide by zero

Reply:
More divide by zero potential cases:

- https://github.com/alexanderjyuen/navigation2/blob/route_server_corner_smoothing/nav2_route/include/nav2_route/corner_smoothing.hpp#L66
- https://github.com/alexanderjyuen/navigation2/blob/route_server_corner_smoothing/nav2_route/include/nav2_route/corner_smoothing.hpp#L45
- https://github.com/alexanderjyuen/navigation2/blob/route_server_corner_smoothing/nav2_route/include/nav2_route/corner_smoothing.hpp#L48-L51 (for the numerators, if start/center or end/center are zero in both XY to make the sum in `bisector_x` and `bisector_y` zero)